pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -0,0 +1,386 @@
+		"toProcess %v toPrepare %v done %v", cluster.Name, filtered, maxAllowed, unavailable, upgrading, len(notReadyMap), keys(toProcessMap), keys(toPrepareMap), keys(upgradedMap))
+
+	if unavailable > maxAllowed {
+		return fmt.Errorf("cluster [%s] worker-upgrade: more than allowed nodes upgrading for cluster: unavailable %v maxUnavailable %v", clusterName, unavailable, maxAllowed)
+	}
+
+	for name, node := range upgradedMap {
+		if v3.NodeConditionUpgraded.IsTrue(node) {
+			continue
+		}
+
+		if err := uh.updateNodeActive(node); err != nil {
+			return err
+		}
+
+		logrus.Infof("cluster [%s] worker-upgrade: updated node [%s] to uncordon", clusterName, name)
+	}
+
+	for _, node := range toProcessMap {
+		if node.Status.NodePlan.Version == cluster.Status.NodeVersion {
+			continue
+		}
+
+		if err := uh.processNode(node, cluster, "upgrading"); err != nil {
+			return err
+		}
+
+		logrus.Infof("cluster [%s] worker-upgrade: updated node [%s] to upgrade", clusterName, node.Name)
+	}
+
+	toDrain := cluster.Spec.RancherKubernetesEngineConfig.UpgradeStrategy.Drain
+	var nodeDrainInput *v3.NodeDrainInput
+	state := "cordon"
+	if toDrain {
+		nodeDrainInput = cluster.Spec.RancherKubernetesEngineConfig.UpgradeStrategy.DrainInput
+		state = "drain"
+	}
+
+	for _, node := range toPrepareMap {
+		if unavailable == maxAllowed {
+			break
+		}
+		unavailable++
+
+		if err := uh.prepareNode(node, toDrain, nodeDrainInput); err != nil {
+			return err
+		}
+
+		logrus.Infof("cluster [%s] worker-upgrade: updated node [%s] to %s", clusterName, node.Name, state)
+	}
+
+	if done == filtered {
+		logrus.Debugf("cluster [%s] worker-upgrade: cluster is done upgrading, done %v len(nodes) %v", clusterName, done, filtered)
+		if !v3.ClusterConditionUpgraded.IsTrue(cluster) {
+			clusterCopy := cluster.DeepCopy()
+			v3.ClusterConditionUpgraded.True(clusterCopy)
+			v3.ClusterConditionUpgraded.Message(clusterCopy, "")
+
+			if _, err := uh.clusters.Update(clusterCopy); err != nil {
+				return err
+			}
+
+			logrus.Infof("cluster [%s] worker-upgrade: finished upgrade", clusterName)
+		}
+	}
+
+	return nil
+}
+
+func (uh *upgradeHandler) toUpgradeCluster(cluster *v3.Cluster) (bool, error) {
+	if v3.ClusterConditionUpgraded.IsUnknown(cluster) {
+		return true, nil
+	}
+
+	nodes, err := uh.nodeLister.List(cluster.Name, labels.Everything())
+	if err != nil {
+		return false, err
+	}
+
+	for _, node := range nodes {
+		if node.Status.NodeConfig == nil {
+			continue
+		}
+
+		if !workerOnly(node.Status.NodeConfig.Role) {
+			continue
+		}
+
+		if node.Status.NodePlan == nil {
+			continue
+		}
+
+		nodePlan, err := uh.getNodePlan(node, cluster)
+		if err != nil {
+			return false, err
+		}
+
+		if planChangedForUpgrade(nodePlan, node.Status.NodePlan.Plan) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func getNum(maxUnavailable string, nodes int) (int, error) {
```

Comment:
Can we rename this function to describe what we are getting? I guess this is similar to `CalculateMaxUnavailable` in RKE?